tests/unit/tf/models/test_retrieval.py

```diff
@@ -997,3 +997,51 @@ def test_youtube_dnn_topk_evaluation(sequence_testing_data: Dataset, run_eagerly
 
     metrics = topk_model.evaluate(dataloader, return_dict=True, pre=predict_next)
     assert all([metric >= 0 for metric in metrics.values()])
+
+
+@pytest.mark.parametrize(
+    "pos_class_weight", [1, tf.random.uniform(shape=(1000,)), "user_item_intentions"]
+)
+@pytest.mark.parametrize("neg_class_weight", [0.1, tf.random.uniform(shape=(1000,))])
+def test_two_tower_v2_with_contrastive_sample_weight(
+    ecommerce_data: Dataset, pos_class_weight, neg_class_weight
+):
+    from merlin.models.tf.experimental.sample_weight import ContrastiveSampleWeight
+
+    tower_dim = 64
+    data = ecommerce_data
+    input_features = ["user_categories", "item_id"]
+    if isinstance(pos_class_weight, str):
+        input_features += [pos_class_weight]
+    data.schema = data.schema.select_by_name(input_features)
+
+    user_schema = data.schema.select_by_tag(Tags.USER)
+    user_inputs = mm.InputBlockV2(user_schema)
+    query = mm.Encoder(user_inputs, mm.MLPBlock([128, tower_dim]))
+
+    item_schema = data.schema.select_by_tag(Tags.ITEM)
+    item_inputs = mm.InputBlockV2(item_schema)
+    candidate = mm.Encoder(item_inputs, mm.MLPBlock([128, tower_dim]))
+
+    output = mm.ContrastiveOutput(
+        DotProduct(),
+        post=None,
+        schema=data.schema.select_by_tag(Tags.ITEM_ID),
+        negative_samplers="in-batch",
+        store_negative_ids=True,
+    )
+    model = mm.TwoTowerModelV2(query, candidate, outputs=output, schema=data.schema)
+    model.compile(run_eagerly=False, loss="binary_crossentropy")
```

Comment:
I suggest adding `weighted_metrics` here, and testing later if the metrics are the same, but `weighed_metrics` are different when setting `sample_weight`.
As a reference, I do some tests like that [here](https://github.com/NVIDIA-Merlin/models/blob/a0d1d455b34b354f19eadec9536c0870abdc077a/tests/unit/tf/outputs/test_block.py#L302-L514)

Reply:
I added tests of `weighted_metrics`, and observed that top-k metrics were inconsistent when passing 2-d sample weights to the `compute_metrics`. In fact, the 2-d sample weights are used to differentiate between positive and sampled negative candidates of each row. However, the MM ranking metrics are computed with respect to the position of the positive sample in the sorted top-k list. So the information needed by weighted metrics is the weight of the positive candidate only.  
The proposed solution is to extract the positive weights from the 2-d matrix (by convention the positive weights are stored in the first column of this matrix). 
So I added the function `_extract_positive_sample_weights` to `test_step` and `train_step` right before calling `compute_metrics`. Curious to have your thoughts about the proposed function?